docs/coding-guidelines/interop-guidelines.md

```diff
@@ -204,3 +188,273 @@ Guidelines for shim C++ API:
      - Note that sizeof(char) == 1 is guaranteed.
   - Do not use size_t in shim API. Always pick a fixed size. Often, it is most convenient to line up with the managed int as int32_t (e.g. scratch buffer size for read/write), but sometimes we need to handle huge sizes (e.g. memory mapped files) and therefore use uint64_t.
   - Use int64_t for native off_t values.
+
+# P/Invoke guidelines
+
+These are guidelines to help make sure your pinvokes are correct and performant.
+
+Attributes
+----------
+
+**[PREFER]** [In] and [Out] only when they differ from the implicit behavior.
+
+**Implicit attributes applied to parameter and return values:**
+
+|                  | Implicit Attribute |
+|------------------|--------------------|
+| parameter        | `[In]`             |
+| `out` parameter  | `[Out]`            |
+| `ref` parameter  | `[In],[Out]`       |
+| return value     | `[Out]`            |
+
+**`[DllImport()]` [1] attribute settings:**
+
+| Setting | Recommendation | Details |
+|---------|----------------|---------|
+| [`PreserveSig`][2]   | keep default       | When this is explicitly set to false (the default is true), failed HRESULT return values will be turned into Exceptions (and the return value in the definition becomes null as a result).|
+| [`SetLastError`][3]  | as per API         | Set this to true (default is false) if the API uses GetLastError and use Marshal.GetLastWin32Error to get the value. If the API sets a condition that says it has an error, get the error before making other calls to avoid inadvertently having it overwritten.|
+| [`ExactSpelling`][4] | `true`             | Set this to true (default is false) and gain a slight perf benefit on Windows as the framework will avoid looking for an "A" or "W" version. (See NDirectMethodDesc::FindEntryPoint).|
+| [`CharSet`][5]       | Explicitly use `CharSet.Unicode` or `CharSet.Ansi` when strings are present in the definition | This specifies marshalling behavior of strings and what `ExactSpelling` does when `false`. Be explicit with this one as the documented default is `CharSet.Ansi`. Note that `CharSet.Ansi` is actually UTF8 on Unix (`CharSet.Utf8` is coming). _Most_ of the time Windows uses Unicode while Unix uses UTF8. |
+
+[1]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.aspx
+[2]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.preservesig.aspx
+[3]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.setlasterror.aspx
+[4]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.exactspelling.aspx
+[5]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.dllimportattribute.charset.aspx
+
+Strings
+-------
+
```

Comment:
All/most of these guidelines were folded into the official docs. For example, this section on string is exactly same as  https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#string-parameters (module a few copy edits).

Should we just link to the official docs, and only have dotnet/runtime add-ons (if there are any)?

Reply:
Ha - I hadn't realized how much was the same. @jkoritzinsky looks like you did this in https://github.com/dotnet/docs/pull/9839. Is it obvious to you which parts are specific to us and should be retained?

I suggest we still merge this as a net improvement.

Reply:
All useful content from `interop-pinvokes.md` was moved over. The differences between the two documents that I can see are:

- The doc on docs.microsoft.com has a few extra details, e.g. about fixed buffers
- The doc on docs.microsoft.com has references to internal implementation details (e.g. names of methods in the C++ runtime impl) deleted.

Reply:
Alright. Shall I just simply delete interop-pinvokes.md and just remove link from interop-guidelines.md ?

Reply:
>Shall I just simply delete interop-pinvokes.md and just remove link from interop-guidelines.md ?

I think so. Our internal guidelines should almost always be the official "best-practices".

Reply:
You could link to the best practices url?

Reply:
OK, done.